contract/src/main.rs

```diff
@@ -1271,206 +1277,353 @@ pub extern "C" fn set_token_metadata() {
+    let burn = EntryPoint::new(
+        ENTRY_POINT_BURN,
+        vec![],
+        CLType::Unit,
+        EntryPointAccess::Public,
+        EntryPointType::Contract,
+    );
+
+    // This entrypoint transfers ownership of token from one account to another.
+    // It looks up the owner of the supplied token_id arg. Revert if token is already burnt,
+    // token_id is invalid, or if caller is not owner and not approved operator.
+    // If token id is invalid it reverts with error InvalidTokenID.
+    let transfer = EntryPoint::new(
+        ENTRY_POINT_TRANSFER,
+        vec![
+            Parameter::new(ARG_SOURCE_KEY, CLType::Key),
+            Parameter::new(ARG_TARGET_KEY, CLType::Key),
+        ],
+        CLType::Tuple2([Box::new(CLType::String), Box::new(CLType::Key)]),
+        EntryPointAccess::Public,
+        EntryPointType::Contract,
+    );
+
+    // This entrypoint approves another account (an operator) to transfer tokens. It reverts
+    // if token_id is invalid, if caller is not the owner, if token has already
+    // been burnt, or if caller tries to approve themselves as an operator.
+    let approve = EntryPoint::new(
+        ENTRY_POINT_APPROVE,
+        vec![Parameter::new(ARG_OPERATOR, CLType::Key)],
+        CLType::Unit,
+        EntryPointAccess::Public,
+        EntryPointType::Contract,
+    );
+
+    let set_approval_for_all = EntryPoint::new(
+        ENTRY_POINT_SET_APPROVE_FOR_ALL,
+        vec![
+            Parameter::new(ARG_TOKEN_OWNER, CLType::Key),
+            Parameter::new(ARG_APPROVE_ALL, CLType::Bool),
+            Parameter::new(ARG_OPERATOR, CLType::Key),
+        ],
+        CLType::Unit,
+        EntryPointAccess::Public,
+        EntryPointType::Contract,
+    );
+
+    // This entrypoint returns the token owner given a token_id. It reverts if token_id
+    // is invalid. A burnt token still has an associated owner.
+    let owner_of = EntryPoint::new(
+        ENTRY_POINT_OWNER_OF,
+        vec![],
+        CLType::Key,
+        EntryPointAccess::Public,
+        EntryPointType::Contract,
+    );
+
+    // This entrypoint returns the operator (if any) associated with the provided token_id
+    // Reverts if token has been burnt.
+    let get_approved = EntryPoint::new(
+        ENTRY_POINT_GET_APPROVED,
+        vec![],
+        CLType::Option(Box::new(CLType::Key)),
+        EntryPointAccess::Public,
+        EntryPointType::Contract,
+    );
+
+    // This entrypoint returns number of owned tokens associated with the provided account
+    let balance_of = EntryPoint::new(
+        ENTRY_POINT_BALANCE_OF,
+        vec![Parameter::new(ARG_TOKEN_OWNER, CLType::Key)],
+        CLType::U64,
+        EntryPointAccess::Public,
+        EntryPointType::Contract,
+    );
+
+    // This entrypoint returns the metadata associated with the provided token_id
+    let metadata = EntryPoint::new(
+        ENTRY_POINT_METADATA,
+        vec![],
```

Comment:
vec![], // <- either HASH or INDEX

Reply:
(drop this in on all the args that are like this)

Reply:
Added in places where necessary